_episodes/02-basic-concepts.md

```diff
@@ -208,4 +233,239 @@ thread hierarchy and the meaning of each of the two arguments,
+
+    return(EXIT_SUCCESS);
+}
+~~~
+{: .language-cuda}
+
+
+### 3.2 Dealing with errors and kernel launches
+
+There are numerous problems that could arise when using CUDA kernels. Some problems may not be related to any custom code
+but to hardware or software incomptability problems. For instance, if running
+the compiled `hello.cu` binary file did not yield any stdout from the GPU, then this might be
+due to an incomptabile driver. To uncover such problems, we need to further improve the code in `hello.cu`.
+
+Unfortunately, CUDA kernel launches do not return any error code, so we will need to make use of the runtime API functions such as
+`cudaPeekAtLastError()` and `cudaGetLastError ` to catch any prelaunch errors. Both functions are similar in functionality in returning the last error that has been produced by any of the runtime calls in the same host thread, but the latter does the additional step of resetting the error to `cudaSuccess`. For the sake of demonstration, we will only need `cudaPeekAtLastError` here.
+
+Since kernel launches are asynchronous, the application must 1st synchronize inbetween the kernel launch and then call cudaPeekAtLastError(), 
+which we achieve by calling another runtime API function: `cudaDeviceSynchronize`. Synchronisation is not always necessary if it is followed by a blocking API call. In this case, we will synchornize all threads between the host and the device, and then exit. The modified code is shown below.
+
+~~~
+/**********************************************/
+
+int main(int argc, char **argv) {
+    helloFromCPU();                       /* Calling from host */
+
+    helloFromGPU<<<1, 8>>>();             /* Calling from device */
+
+    checkCudaErrors( cudaPeekAtLastError() );
+    checkCudaErrors( cudaDeviceSynchronize() );    /* Force the host to wait for the device to check for error */ 
+
+    /* Do more stuff ... */
+
+    checkCudaErrors( cudaDeviceReset() );         /* House-keeping on device */
+
+    return(EXIT_SUCCESS);
+}
+~~~
+{: .language-cuda}
+
+We can compile this code just like before. A common problem encountered when setting up CUDA infrastructure for the 1st is driver incomptability. 
+If we run `hello.cu` without checking for kernel errors, we get stdout only from the CPU. With error checking for kernel launches taken into account, we can
+decipher the problem by catching the runtime error:
+~~~
+$ ./hello
+~~~
+{: .language-bash}
+
+will print the following output
+
+~~~
+Hello World from CPU!
+GPUassert: CUDA driver version is insufficient for CUDA runtime version hello.cu 32
+~~~
+{: .output}
+
+In this case, the runtime error has informed us that we need to install a newer driver for successful code execution
+on the device. This is a typical error beginners encounter on the GNU/Linux OS.
+
+For more info on error handling, see the more detailed [Error Handling](/04-gpu-compilation-model#3-error-handling) section.
```

Comment:
I intend to keep the error handling brief as is within Compilation Model in lesson 3. It can be included in Lesson 2 as a basic concept here, but we need to find a way to bind it to the rest of the story within the context. At the moment, seems out of place to me. If we choose the former option to keep it in Lesson 3, we can move these and merge them together without disturbing the whole tutorial structure which I am trying to avoid.

Reply:
IMO this should be covered right from the beginning and any code u introduce later on should make use of error checking, because this is a good practice so u wouldn't wanna cover it in the end.

Reply:
I would be less flexible here but on the intermediate and advanced tutorials where we do not have any materials laid out yet, I will be much more open to anything you are going to add there especially if it is already in place like this one.